connect/cli/plugins/project/extension/validations.py

```diff
@@ -234,64 +269,173 @@ def validate_schedulables(config, project_dir, context):
-                ),
-            )
-            continue
+    return ValidationResult(messages, False, context)
 
-        if variable["name"] in names:
-            messages.append(
-                ValidationItem(
-                    'ERROR',
-                    f'Duplicate variable name: the variable with name *{variable["name"]}* '
-                    'has already been declared.',
-                    **get_code_context(extension_class, '@variables'),
-                ),
-            )
 
-        names.append(variable["name"])
+def validate_webapp_extension(config, project_dir, context):  # noqa: CCR001
 
-        if not variable_name_regex.match(variable['name']):
-            messages.append(
-                ValidationItem(
-                    'ERROR',
-                    f'Invalid variable name: the value *{variable["name"]}* '
-                    f'does not match the pattern *{variable_name_pattern}*.',
-                    **get_code_context(extension_class, '@variables'),
-                ),
-            )
-        if 'initial_value' in variable and not isinstance(variable['initial_value'], str):
-            messages.append(
-                ValidationItem(
-                    'ERROR',
-                    f'Invalid *initial_value* attribute for variable *{variable["name"]}*: '
-                    f'must be a non-null string not *{type(variable["initial_value"])}*.',
-                    **get_code_context(extension_class, '@variables'),
-                ),
-            )
+    messages = []
+
+    if 'webapp' not in context['extension_classes']:
+        return ValidationResult(messages, False, context)
+
+    extension_class = context['extension_classes']['webapp']
+    extension_class_file = inspect.getsourcefile(extension_class)
+
+    if not inspect.getsource(extension_class).strip().startswith('@web_app(router)'):
+        messages.append(
+            ValidationItem(
+                'ERROR',
+                'The Web app extension class must be wrapped in *@web_app(router)*.',
+                extension_class_file,
+            ),
+        )
+        return ValidationResult(messages, True)
 
-        if 'secure' in variable and not isinstance(variable['secure'], bool):
+    has_router_function = False
+    for _, value in inspect.getmembers(extension_class):
+        if (
+                inspect.isfunction(value)
+                and inspect.getsource(value).strip().startswith('@router.')
+        ):
+            has_router_function = True
+            break
+
+    if not has_router_function:
+        messages.append(
+            ValidationItem(
+                'ERROR',
+                'The Web app extension class must contain at least one router '
+                'implementation function wrapped in *@router.your_method("/your_path")*.',
+                extension_class_file,
+            ),
+        )
+        return ValidationResult(messages, True)
+
+    if 'ui' not in context['descriptor']:
```

Comment:
The ui field should not be mandatory. An extension could have a class inside webapp.py and do not have static files. What I suggest is to validate the static files ONLY if ui exists on the descriptor (what you do on lines 390-426).

Reply:
that's correct, an extension can provide api and not UI but is same type at the end